### DISCUSSION TITLE
Adding hydrogens to a conformer based on template mol

### DISCUSSION THREAD

**Discussion by YoavShamir5** [rdkit/rdkit#8117] [Q&A]
Hello, I'd appreciate advice on a technical question -  I have two RDkit molecules loaded from files with 3D coordinates (different coordinates for each mol) - mol A has no hydrogens (green), and mol B has hydrogens with coordinates (cyan):
![image](https://github.com/user-attachments/assets/c047acd5-797c-4035-9b29-c047385dd1bf)

Both compounds are chemically identical, apart from the presence/absence of hydrogens. I would like to add hydrogens (with 3D coordinates) to mol A according to the template molecule, mol B. Simply adding hydrogens in Pymol, for example, would not do the trick - for example, Pymol would not add a hydrogen attached to the orange nitrogen, as I want it to. Breaking it to two parts:
1) Does Rdkit enable the mapping of hydrogens from a template (mol B) to another mol (mol A)?
2) If 1 is doable - can I model the 3D coordinates of these added hydrogens on mol B, keeping all other atoms of mol B (non-hydrogens) at their existing coordinates? Thanks!

**Comment by ptosco** [comment]
You may find the heavy atom mappings between A and B through a substructure match. Once you have mapped heavy atoms, get the formal charge for each heavy atom on mol B and set it for the matching atom on mol A. Then call `Chem.AddHs(molB, addCoords=True)`.

**Reply by YoavShamir5** [reply]
Thanks for your help, @ptosco !

I am trying this out, and am not sure why, but when I use the pattern molecule directly read from an SDF file, the substructure match fails. When I convert it to SMILES and then back to an rdkit mol, it works:
```
supplier = Chem.SDMolSupplier(output_file,sanitize=False)
mol_from_sdf = [mol for mol in supplier][0]
sdf_mol_smi = Chem.MolToSmiles(mol_from_sdf)
print("sdf_mol_smi: ", sdf_mol_smi)
mol_smi = "Cc1ccccc1Nc1nc2cc(F)c(N(C)C(=O)CC([SiH3])C[NH+](C)C)cc2n2cncc12"
mol = Chem.MolFromSmiles(mol_smi)

matches_a = mol.GetSubstructMatches(mol_from_sdf)  #this fails to find matches
print("matches_a:", matches_a)

sdf_mol_new = Chem.MolFromSmiles(sdf_mol_smi)
print("sdf_mol_new: ", Chem.MolToSmiles(sdf_mol_new))
matches_b = mol.GetSubstructMatches(sdf_mol_new)  #this works ok
print("matches_b:", matches_b)
sdf_mol_smi:  CC1=CC=CC=C1N/C1=N/C2=CC(F)=C(N(C)C(=O)CCCN(C)C)C=C2N2C=NC=C12
matches_a: ()
sdf_mol_new:  Cc1ccccc1Nc1nc2cc(F)c(N(C)C(=O)CCCN(C)C)cc2n2cncc12
matches_b: ((0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32),)
```

What function am I forgetting to run, which will replace this mol-SMILES-mol conversion? 
Also - how can I use the match for mapping formal charges of the heavy atoms of both molecules? Thanks!

**Reply by YoavShamir5** [reply]
I fixed my above issue by kekulizing both molecules, but at the next step you suggested, addHs does not add any atoms:

```
print(mol_from_sdf.GetNumAtoms())
print(Chem.MolToSmiles(mol_from_sdf))

mol_from_sdf_with_h = Chem.AddHs(mol_from_sdf, addCoords=True)

print(Chem.MolToSmiles(mol_from_sdf_with_h))
print(mol_from_sdf_with_h.GetNumAtoms())
32
[C]c1[c][c][c][c]c1[N]c1nc2[c]c(F)c(N([C])C(=O)[C][C][C][N+]([C])[C])[c]c2n2[c]n[c]c12
[C]c1[c][c][c][c]c1[N]c1nc2[c]c(F)c(N([C])C(=O)[C][C][C][N+]([C])[C])[c]c2n2[c]n[c]c12
32
```

Am I missing some function here? Thanks!

**Comment by ptosco** [comment]
@YoavShamir5 Could you please attach your SDF file? You will only need to change its extension from `.sdf` to `.txt`.

**Comment by YoavShamir5** [comment]
Hello @ptosco, thank you for your reply, I've managed to reduce the issue to a substructure match, that seems to works for some of my sdf molecules, but not for others. Here is the code: 

```
test_mol_from_sdf = [mol for mol in supplier][0]

#test_mol_from_sdf_smi = Chem.MolToSmiles(test_mol_from_sdf)  #returns: CCC(=O)NC1=C(OC)C=CC(NC(=O)NC(C)(C)C2=CC(C(F)(F)F)=CC=C2)=C1
#test_mol_from_sdf = Chem.MolFromSmiles(test_mol_from_sdf_smi)
 
orig_mol = Chem.MolFromSmiles("COc1ccc(NC(=O)[N-]C(C)(C)c2cccc(C(F)(F)F)c2)cc1NC(=O)CC[SiH3]")
matches = orig_mol.GetSubstructMatches(test_mol_from_sdf)  

print(matches)
```

With those two lines commented out, there is not match. But If uncomment, the single match is found. 
I would like to use the SDF molecules directly (I use its coordinates), what should be tweaked to get these to to match without the commented-out lines? Thanks! (attached is the sdf)

[m1000_model_lig.txt](https://github.com/user-attachments/files/18252291/m1000_model_lig.txt)

**Reply by YoavShamir5** [reply]
This seems to be related to aromaticity and kekulization - is there a straightforward way to conduct substructure matching ignoring all of the bond types? I have molecule pairs that should definitely match, up to bond type.

**Comment by ptosco** [comment]
@YoavShamir5 
The molecule you attached, [m1000_model_lig.txt](https://github.com/user-attachments/files/18252291/m1000_model_lig.txt), has no explicit hydrogens.
Instead, it has nonzero valences in the atom block valence field, which are therefore filled by RDKit with radical electrons.
You may see this when you print out the SMILES string corresponding to your input molecule:
```
from rdkit import Chem

mb = """m1000
 OpenBabel12262412153D

 30 31  0  0  0  0  0  0  0  0999 V2000
   12.0790   -8.2210   -1.0020 C   0  0  0  0  0  1  0  0  0  0  0  0
   11.5050   -6.9440   -1.0360 O   0  0  0  0  0  0  0  0  0  0  0  0
   11.2970   -6.2580    0.0970 C   0  0  0  0  0  0  0  0  0  0  0  0
   11.5510   -6.7870    1.3670 C   0  0  0  0  0  3  0  0  0  0  0  0
   11.3460   -5.9730    2.5130 C   0  0  0  0  0  3  0  0  0  0  0  0
   10.9280   -4.6630    2.3710 C   0  0  0  0  0  0  0  0  0  0  0  0
   10.7780   -3.7720    3.4020 N   0  0  0  0  0  2  0  0  0  0  0  0
   10.1460   -3.9790    4.5680 C   0  0  0  0  0  0  0  0  0  0  0  0
    9.8100   -5.0630    4.9400 O   0  0  0  0  0  0  0  0  0  0  0  0
    9.9530   -2.8060    5.2590 N   0  0  0  0  0  2  0  0  0  0  0  0
    9.0310   -2.4900    6.3160 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.7910   -1.9140    5.6410 C   0  0  0  0  0  1  0  0  0  0  0  0
    8.6440   -3.7160    7.1330 C   0  0  0  0  0  1  0  0  0  0  0  0
    9.6540   -1.4620    7.2530 C   0  0  0  0  0  0  0  0  0  0  0  0
   10.6880   -1.8330    8.1180 C   0  0  0  0  0  3  0  0  0  0  0  0
   11.2170   -0.8970    8.9780 C   0  0  0  0  0  3  0  0  0  0  0  0
   10.7400    0.4100    8.9880 C   0  0  0  0  0  3  0  0  0  0  0  0
    9.7020    0.7540    8.1370 C   0  0  0  0  0  0  0  0  0  0  0  0
    9.1270    2.1630    8.1530 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.0560    2.2020    8.9360 F   0  0  0  0  0  0  0  0  0  0  0  0
    8.7730    2.5400    6.9430 F   0  0  0  0  0  0  0  0  0  0  0  0
    9.9880    3.0560    8.6150 F   0  0  0  0  0  0  0  0  0  0  0  0
    9.1750   -0.1720    7.2690 C   0  0  0  0  0  3  0  0  0  0  0  0
   10.6670   -4.1590    1.0950 C   0  0  0  0  0  3  0  0  0  0  0  0
   10.7950   -4.9330   -0.0250 C   0  0  0  0  0  0  0  0  0  0  0  0
   10.4500   -4.4270   -1.3070 N   0  0  0  0  0  2  0  0  0  0  0  0
    9.5990   -3.3470   -1.5280 C   0  0  0  0  0  0  0  0  0  0  0  0
    9.1790   -2.6620   -0.6120 O   0  0  0  0  0  0  0  0  0  0  0  0
    9.2470   -3.0500   -2.9520 C   0  0  0  0  0  2  0  0  0  0  0  0
    7.8550   -2.9520   -3.1630 C   0  0  0  0  0  1  0  0  0  0  0  0
  2  1  1  0  0  0  0
  2  3  1  0  0  0  0
  3  4  1  0  0  0  0
  4  5  2  0  0  0  0
  6  5  1  0  0  0  0
  6  7  1  0  0  0  0
  7  8  1  0  0  0  0
  8  9  2  0  0  0  0
  8 10  1  0  0  0  0
 10 11  1  0  0  0  0
 11 13  1  0  0  0  0
 11 14  1  0  0  0  0
 12 11  1  0  0  0  0
 14 23  2  0  0  0  0
 14 15  1  0  0  0  0
 15 16  2  0  0  0  0
 16 17  1  0  0  0  0
 18 19  1  0  0  0  0
 18 17  2  0  0  0  0
 19 22  1  0  0  0  0
 19 20  1  0  0  0  0
 21 19  1  0  0  0  0
 23 18  1  0  0  0  0
 24  6  2  0  0  0  0
 25  3  2  0  0  0  0
 25 24  1  0  0  0  0
 26 25  1  0  0  0  0
 27 26  1  0  0  0  0
 27 28  2  0  0  0  0
 29 27  1  0  0  0  0
 30 29  1  0  0  0  0
M  END
$$$$
"""
supplier = Chem.SDMolSupplier()
supplier.SetData(mb)
test_mol_from_sdf = next(supplier)
print(Chem.MolToSmiles(test_mol_from_sdf))
[C][C]C(=O)[N]c1[c]c([N]C(=O)[N]C([C])([C])c2[c][c][c]c(C(F)(F)F)[c]2)[c][c]c1O[C]
```
This is the reason why the substructure match fails:
```
orig_mol = Chem.MolFromSmiles("COc1ccc(NC(=O)[N-]C(C)(C)c2cccc(C(F)(F)F)c2)cc1NC(=O)CC[SiH3]")
matches = orig_mol.GetSubstructMatches(test_mol_from_sdf)
print(matches)
()
```
Since all valences are already filled by radicals, also attempting to add explicit Hs has no effect:
```
test_mol_from_sdf_with_h = Chem.AddHs(test_mol_from_sdf, addCoords=True)
print(test_mol_from_sdf_with_h.GetNumAtoms())
30
```
If you convert the radicals into implicit Hs, the SMILES will not show absence of implicit Hs anymore, the substructure match will succeed, and also explicit hydrogens will be added as expected:
```
for a in test_mol_from_sdf.GetAtoms():
    rad = a.GetNumRadicalElectrons()
    if rad:
        a.SetNumExplicitHs(rad)
        a.SetNumRadicalElectrons(0)
print(Chem.MolToSmiles(test_mol_from_sdf))
CCC(=O)Nc1cc(NC(=O)NC(C)(C)c2cccc(C(F)(F)F)c2)ccc1OC
orig_mol = Chem.MolFromSmiles("COc1ccc(NC(=O)[N-]C(C)(C)c2cccc(C(F)(F)F)c2)cc1NC(=O)CC[SiH3]")
matches = orig_mol.GetSubstructMatches(test_mol_from_sdf)
print(matches)
((0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29),)
test_mol_from_sdf_with_h = Chem.AddHs(test_mol_from_sdf, addCoords=True)
print(test_mol_from_sdf_with_h.GetNumAtoms())
54
```
You should probably investigate why hydrogens were stripped from your 3D file and replaced with radicals in the first place.
